spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -385,16 +387,40 @@ private static void mergeIcebergHadoopConfs(
         }
       }
 
-      try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
-        this.tasks = Lists.newArrayList(tasksIterable);
-      } catch (IOException e) {
-        throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
-      }
+      this.tasks = planScan(scan);
```

Comment:
What about making `PlanScanAction` implement `TableScan` instead of passing it a `TableScan`? Then you'd instantiate it above in place of `scan` and configure it the same way.

Reply:
This is a pretty fun Idea, let me try that

Reply:
The awkward thing about this is that `TableScan` is an API based on refinement, so you'd be creating a new action each time, just like we do in `BaseTableScan`. That's not terrible, but it is a bit of a surprise since the other actions use a builder pattern instead.

Reply:
One awkward thing I've just hit as I've just about finished this refactor, If we make the decision when we are building the scan and not when we call planFiles, we can't fallback from distributed to not distributed if it doesn't work, unless we maintain two scan objects